### DISCUSSION TITLE
Sleep Moddable Two

### DISCUSSION THREAD

**Discussion by mauroForlimpopoli** [Moddable-OpenSource/moddable#589] [General]
Is there a way to put the moddable two in sleep mode ?

**Comment by mauroForlimpopoli** [comment]
I resolved with a deepSleep.c file with this funcion
#include  "xs.h"

void  xs_system_deepSleep (xsMachine * the) 
{ 
	esp_deep_sleep_start();
}.

But now how can I do that during sleep it run my "RTC function" ?

**Reply by phoddie** [reply]
Nice. You answered your own question about deep sleep.

Your question about the RTC is more of an ESP32 question than a Moddable SDK question. Deep sleep is particularly challenging because the content of memory is lost. That means your program stops running and is restarted when the device wakes. [This tutorial](https://www.instructables.com/ESP32-Deep-Sleep-Tutorial/) over on Instructables looks like a good starting point.

**Reply by mauroForlimpopoli** [reply]
The Sleep and Wake functions work correctly. I have the problen  about to put my function into RTC memory.
The example I found:
void RTC_IRAM_ATTR esp_wake_deep_sleep(void) {
    esp_default_wake_deep_sleep();
    // Add additional functionality here
}
and the attribute of:
RTC_DATA_ATTR int wake_count;
give error in the compiler.
The RTC data memory is located 
0x3FF8_0000 to 0x3FF8_1FFF  
0x400C_0000 to 0x400C_1FFF 
for a total of 8k
Is it possible to choose where write my function/variables ?
Thank you

**Comment by phoddie** [comment]
@mauroForlimpopoli - when you report a problem (e.g. "give error in the compiler"), it would help if you could provide the error message and the complete example.

My guess is that your didn't include the necessary header file, so you got an error like this:

```
~/Projects/moddable/examples/helloworld/restart.c:5:14: error: expected ';' before 'int'
 RTC_DATA_ATTR int wake_count;
```

This implementation of `deepSleep` compiles. It includes the `esp_sleep.h`. I've added support for specifying a wake time, as in the Instructables example above. It also includes a `wakeCount()` function that returns the RTC memory counter and increments it.

```c
#include "xs.h"
#include "esp_sleep.h"

RTC_DATA_ATTR int wake_count;

void xs_deepSleep(xsMachine * the)
{
	esp_sleep_enable_timer_wakeup(xsToInteger(xsArg(0)) * 1000);
	esp_deep_sleep_start();
}

void xs_wakeCount(xsMachine * the)
{
	xsResult = xsInteger(wake_count);
	wake_count += 1;
}
```

Here's a JavaScript example that deep sleeps for 500 seconds intervals and outputs the wake count on each wake.

```js
function deepSleep(ms) @ "xs_deepSleep"
function wakeCount() @ "xs_wakeCount"

trace(wakeCount() + "\n");

deepSleep(500);
```

**Reply by mauroForlimpopoli** [reply]
Dear Peter,
it was like you soppose....and now the code works.
You are a wizard  👍 
Thank you again

I wish to create variable who inc of "one" each second during sleep  to build a clock in javascript. So this is the code:
////////////// code
#include "xsmc.h"
#include "stdint.h"
#include "mc.xs.h"		
#include "esp_sleep.h"
#include "unistd.h"

RTC_DATA_ATTR int wake_count;
RTC_DATA_ATTR int lapTime;
void RTC_IRAM_ATTR esp_wake_deep_sleep(void) {

	int i = 0;
	while (i < 100000) {
	 sleep(2000); /////////////////////////dont't work
	 lapTime += 1;
	  i++;
	}
	
}

void xs_deepSleep(xsMachine * the)
{
	esp_sleep_enable_ext0_wakeup(0, 0);
  	esp_deep_sleep_start();

}

void xs_wakeCount(xsMachine * the)
{
	xsResult = xsInteger(lapTime);
	wake_count += 1;
}
//////////////////
But the instruction sleep(2000); don't work.  (I think it could be normal the CPU is in sleep node yet!)
 Have you some suggestions on how to do a wait time of about a second?

**Reply by phoddie** [reply]
You cannot have a function running in deep sleep. The CPU is off. The RTC module continues to run, so you can wake up. The wake operation is a near full reboot of the ESP32.

In addition to retaining the RTC memory, the RTC also maintains the clock in deep sleep. So, you don't need to implement a clock. You can just use the built-in clock through the JavaScript Date object.

**Reply by mauroForlimpopoli** [reply]
Sorry, can you give me more datails ?
I don't understand how can I resume the time from the RTC memory
Thank you

**Reply by phoddie** [reply]
There is no need to resume the time. It is automatically retained by the RTC.

You can display the current time to the console before going to sleep to confirm that:


```js
trace(`The time is ${new Date}\n`);
```

**Reply by mauroForlimpopoli** [reply]
Sorry again, but if I declare a variable ie:  let d = new Date(); in the main, when the device exit from sleep, overwrites the variable, only the variables writen in RTC are not overwritten. Where I wrong ? (I forget to tell that che device is not conneted to a network!)

**Reply by phoddie** [reply]
You are correct. When you put the ESP32 into deep sleep, all memory is lost. So, the variables are lost. When the device wakes, it begins running the software again. That does not overwrite the previous values, since they are already lost on entry to deep sleep.

For any variables you want to persist between deep sleep cycles, you need to take some additional steps in your code. Here are two options:

- For frequently updated variables, declare additional RTC  backed variables in your C code and additional accessor functions, like `wakeCount()` above, to read and write them.

```c
RTC_DATA_ATTR int anotherVariable;

void xs_getAnotherVariable(xsMachine * the)
{
    xsResult = xsInteger(anotherVariable);
}

void xs_setAnotherVariable(xsMachine * the)
{
    anotherVariable = xsToInteger(xsArg(0)));
}
```
```js
function getAnotherVariable() @ "xs_getAnotherVariable"
function setAnotherVariable() @ "xs_setAnotherVariable"

let anotherVariable = getAnotherVariable();
setAnotherVariable(anotherVariable + 1);
```

- For infrequently updated variables, use a [preference](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/files/files.md#class-preference). These will survive both deep sleep and a complete loss of power.

**Reply by mauroForlimpopoli** [reply]
This is clear, but I need to set a variable who change (+1) every second, in sleep mode too. The ULP Coprocessor works with a frequency of 8mhz, I need a delay of a second and a routine with a "while cycle" with high value it seems don't work.

**Reply by phoddie** [reply]
Deep sleep turns the processor off. It cannot run JavaScript or C code when it is off. 

You can use the ULP while in deep sleep. It has its own assembly language that you can use to write programs. That is out of scope for this project. It is a purely an ESP-IDF matter. I suggest you review the ESP-IDF documentation and examples on the ULP. If you have questions, they are best answered on a forum dedicated to the ESP32.

**Reply by mauroForlimpopoli** [reply]
Ok, Thank you

**Reply by mauroForlimpopoli** [reply]
At least I decide to use a network to get time. 
But I have this strange problem.
This is the timer of my main.
Timer.repeat(() => {
	let d = new Date;
	d= d.toString();
	d= d.slice(4,24)
 timeLabel.string = d;
}, 1000);

this is my connection:

function connectWifi () {
	let wifiMonitor = new WiFi({
	 ssid: "Xperia XZ1_ed16",
	 password: "xxxxxx"
	 },
	 function(msg) {
		 switch (msg) {
		 case WiFi.gotIP:
			 trace("network ready\n");
				 
		 break;
		 case WiFi.connected:
			trace("connected\n");
		 break;
		 case WiFi.disconnected:
			trace("connection lost\n");
		 break;
		 }
	 }
	);
}

It works only if the Board is connected to the USB/serial adapter, if it works alone it doesn't work.
This are my include
import {} from "piu/MC";
import {Keyboard, BACKSPACE, SUBMIT} from "keyboard";
import Timer from "timer";
import Timeline from "piu/Timeline";
import Digital from "pins/digital";
import Analog from "pins/analog";
import PWM from "pins/pwm";
import WiFi from "wifi";
import { Request } from "http";
import WipeTransition from "piu/WipeTransition";
import config from "mc/config";
import {File, Iterator, System} from "file";
import Resource from "Resource";
import I2C from "pins/i2c";
...I need to add something else ?
Thank you

**Reply by phoddie** [reply]
You don't need to do anything special to run on USB power.

You say it doesn't work. How does it fail?

**Reply by mauroForlimpopoli** [reply]
simply a get 01/01/1970  00:00 ...
and the time go on....
after 5 minutes 00:05 I cause a sleep for some seconds......it wake up with 00:05. It keeps the time. Only I don't read the right date when it's wifi connected. With programmer board attached it runs right

**Reply by phoddie** [reply]
There is no battery backed clock on the Moddable Two board. The time needs to be initialized somehow.

When using the xsbug for debugging, the ESP32 clock is automatically initialized by xsbug when the debugging connection is established. When you are not debugging, you need to initialize the time somehow.

Since you are managing the Wi-Fi connection yourself, you will need to initialize the time in your code. This is explained in our  book [IoT Development for ESP32 and ESP8266 with JavaScript](https://www.moddable.com/book) on pages 177 and 178. You might also take a look at the [SNTP example](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/network/sntp/main.js) in the Moddable SDK.

**Reply by mauroForlimpopoli** [reply]
Thank you, It was necessary to set the SNTP  .

**Reply by mauroForlimpopoli** [reply]
Good morning,
when the ESP32 is in "Deep-sleep" mode, the moddable two absorbs about 2mA if supplied on 3.3 Vdc. In the datasheet the consumption  is about 100uA. I suppose the difference could be done cause the supply of other ICs on 3,3 Vdc. Is there a way to disable che TouchScreen IC or LCD IC by software command ?
Thank you

**Reply by mauroForlimpopoli** [reply]
After more accurate test, I misured a current of 8/10 mA (3.3 Vdc). Either with the "Deep Sleep" command or by keeping the "Reset" button pressed. This consumption could be from the ICs of the device? 
The FT6206 and ILI9341 drivers have a "Reset pin" (no. 36 and 6). It seems that by keeping them low, the devices are set in "sleep mode" to reduce the consumption. Are they connected in the board? Where? 
I found this indication to set the RST line of ILI9341 .

```
"defines": {
	"ili9341": {
		"width": 240,
		"height": 320,
		"hz": 10000000,
		"flipX": true,
		"flipY": false,
		...
		"rst": {
			"port": null,
			"pin": 0
		},
		...
```

Is it means that I choose if the reset pin is connected or not ? (How is the default value?)
Thank you

**Comment by phoddie** [comment]
The reset pins of the touch and display controllers are not connected to GPIO pins on Moddable Two.
